libraries/Standard/src/primitive/resample.birch

```diff
@@ -35,20 +35,31 @@ function nan_max(w:Real[_]) -> Real {
 /*
  * Exponentiate and sum a log weight vector.
  *
- * The result is computed in a numerically stable way that avoids
- * intermediate over- and underflow, using a single pass over the data.
- *
  * @param w Log weights.
  *
  * @return the logarithm of the sum.
  *
  * !!! note
  *     NaN log weights are treated as though `-inf`.
+ *
+ * This uses a numerically stable implementation that avoids over- and
+ * underflow, using a single pass over the data.
+ * It is based on:
+ *
+ * S. Nowozin (2016). Streaming Log-sum-exp Computation.
+ * http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html
  */
 function log_sum_exp(w:Real[_]) -> Real {
   if length(w) > 0 {
-    let mx <- -inf; // running maximum of log weights
-    let r <- 0.0; // running sum of non-maximum weights divided by maximum
+    // Running maximum of log weights
+    let mx <- -inf;
+    /*
+     * Running sum of non-maximum weights divided by maximum weight.
+     * In contrast to Nowozin's implementation the maximum itself
+     * is not included in the sum.
+     * This avoids underflow for e.g. w = [1e-20, log(1e-20)].
+     */
+    let r <- 0.0;
     wn:Real;
```

Comment:
Delete, use `let wn <- w[n]` below instead, as `wn` doesn't need to be preserved over iterations of the loop.

Reply:
```suggestion
```